cstrml/staking/src/lib.rs

```diff
@@ -1545,39 +1631,71 @@ impl<T: Trait> Module<T> {
-            let total_rewards = Self::validators(stash).guarantee_fee * reward;
-            let mut guarantee_rewards = Zero::zero();
-
-            for i in &exposure.others {
-                let per_u64 = Perbill::from_rational_approximation(i.value, total);
-                // Reward guarantors
-                guarantee_rewards += per_u64 * total_rewards;
-                imbalance.maybe_subsume(Self::make_payout(&i.who, per_u64 * total_rewards));
-            }
+    fn do_reward_stakers(
+        validator_stash: T::AccountId,
+        era: EraIndex,
+    ) -> DispatchResult {
+        // Validate input data
+        let current_era = CurrentEra::get().ok_or(Error::<T>::InvalidEraToReward)?;
+        ensure!(era <= current_era, Error::<T>::InvalidEraToReward);
+        let history_depth = Self::history_depth();
+        ensure!(era >= current_era.saturating_sub(history_depth), Error::<T>::InvalidEraToReward);
+
+        // Note: if era has no reward to be claimed, era may be future. better not to update
+        // `ledger.claimed_rewards` in this case.
+        let era_staking_payout = <ErasStakingPayout<T>>::get(&era)
+            .ok_or_else(|| Error::<T>::InvalidEraToReward)?;
+
+        let controller = Self::bonded(&validator_stash).ok_or(Error::<T>::NotStash)?;
+        let mut ledger = <Ledger<T>>::get(&controller).ok_or_else(|| Error::<T>::NotController)?;
+
+        ledger.claimed_rewards.retain(|&x| x >= current_era.saturating_sub(history_depth));
+        match ledger.claimed_rewards.binary_search(&era) {
+            Ok(_) => Err(Error::<T>::AlreadyClaimed)?,
+            Err(pos) => ledger.claimed_rewards.insert(pos, era),
+        }
+        /* Input data seems good, no errors allowed after this point */
+        let exposure = <ErasStakers<T>>::get(&era, &ledger.stash);
+        <Ledger<T>>::insert(&controller, &ledger);
+
+        // 1. Pay authoring reward
+        if let Some(value) = <ErasAuthoringPayout<T>>::get(&era, &validator_stash) {
+            if let Some(imbalance) = Self::make_payout(
+                &validator_stash,
+                value
+            ) {
+                Self::deposit_event(RawEvent::Reward(validator_stash, imbalance.peek()));
+            };
+        }
 
-            guarantee_rewards
+        // 2. Pay staking reward
+        let to_num =
+        |b: BalanceOf<T>| <T::CurrencyToVote as Convert<BalanceOf<T>, u64>>::convert(b);
+
+
+        let era_total_stakes = <ErasTotalStakes<T>>::get(&era);
+        let staking_reward = Perbill::from_rational_approximation(to_num(exposure.total), to_num(era_total_stakes)) * era_staking_payout;
+        let total = exposure.total.max(One::one());
+        let total_rewards = <ErasValidatorPrefs<T>>::get(&era, &ledger.stash).guarantee_fee * staking_reward;
+        let mut guarantee_rewards = Zero::zero();
+        for i in &exposure.others {
+            let per_u64 = Perbill::from_rational_approximation(i.value, total);
```

Comment:
`per_u64` seems not reasonable for me😅

Reply:
emm... the type of this value is `Perbill`. The name existed in the old `reward_validators` function and I borrow this code. It's not reasonable for me too. Renamed it to `reward_ratio`

Reply:
cool